File: libchess/piece.py
class Piece(object):
    """Represents a chess piece."""

    def __init__(self, type, color):
        """Inits a piece with type and color.

        Args:
            type: The type of the piece (pawn, bishop, knight, rook, king or
                queen) as "p", "b", "n", "r", "k", or "q".
            color: The color of the piece as "w" for white or "b" for black.
        """
        assert type in ["p", "b", "n", "r", "k", "q"]
        self._type = type
        assert color in ["w", "b"]
        self._color = color

    def get_color(self):
        """Gets the color of the piece.

        Returns:
            "b" for black or "w" for white.
        """
        return self._color

    def get_full_color(self):
        """Gets the full color of the piece.

        Returns:
            "black" or "white".
        """
        if self._color == "w":
            return "white"
        else:
            return "black"

    def get_type(self):
        """Gets the type of the piece.

        Returns:
            "p", "b", "n", "r", "k", or "q" for pawn, bishop, knight, rook
            king or queen.
        """
        return self._type

    def get_full_type(self):
        """Gets the full type of the piece.

        Returns:
            "pawn", "bishop", "knight", "rook", "king" or "queen".
        """
        if self._type == "p":
            return "pawn"
        elif self._type == "b":
            return "bishop"
        elif self._type == "n":
            return "knight"
        elif self._type == "r":
            return "rook"
        elif self._type == "k":
            return "king"
        elif self._type == "q":
            return "queen"

    def get_symbol(self):
        """Gets the symbol representing the piece.

        Returns:
            The symbol of the piece as used in FENs. "p", "b", "n", "r", "k" or
            "q" for a black pawn, bishop, knight, rook, king or queen. "P",
            "B", "N", "R", "K", or "Q" for the corresponding white piece.
        """
        if self._color == "w":
            return self._type.upper()
        else:
            return self._type

    def __str__(self):
        return self.get_symbol()

    def __repr__(self):
        return "Piece.parse_symbol('%s')" % get_symbol()

    def __eq__(self, other):
        return self.get_symbol() == other.get_symbol()

    def __ne__(self, other):
        return self.get_symbol() != other.get_symbol()

    @staticmethod
    def parse_symbol(symbol):
        """Parses a piece symbol.

        Args:
            symbol: The symbol of the piece as used in FENs. "p", "b", "n",
            "r", "k" for a black pawn, bishop, knight, rook or king. "P", "B",
            "N", "R", "K" or "Q" for the corresponding white piece.

        Returns:
            An object of the Piece class.
        """
        type = symbol.lower()
        if type == symbol:
            return Piece(type, "b")
        else:
            return Piece(type, "w")

File: libchess/piece.cpp
#include "piece.h"

#include <boost/format.hpp>
#include <stdlib.h>

namespace chess {

    Piece::Piece(char symbol) {
        switch (tolower(symbol)) {
            case 'p':
            case 'n':
            case 'b':
            case 'r':
            case 'q':
            case 'k':
                m_symbol = symbol;
                break;
            default:
                throw new std::invalid_argument("symbol");
        }
    }

    char Piece::color() const {
        if (m_symbol == toupper(m_symbol)) {
            return 'w';
        } else {
            return 'b';
        }
    }

    char Piece::type() const {
        return tolower(m_symbol);
    }

    char Piece::symbol() const {
        return m_symbol;
    }

    std::string Piece::full_color() const {
        if (color() == 'w') {
            return "white";
        } else {
            return "black";
        }
    }

    std::string Piece::full_type() const {
        switch(type()) {
            case 'p':
                return "pawn";
            case 'b':
                return "bishop";
            case 'n':
                return "knight";
            case 'r':
                return "rook";
            case 'k':
                return "king";
            case 'q':
                 return "queen";
        }
    }

    std::string Piece::__repr__() const {
        return str(boost::format("Piece('%1%')") % m_symbol);
    }

    int Piece::__hash__() const {
        return m_symbol;
    }

    bool Piece::operator==(const Piece& other) const {
        return m_symbol == other.m_symbol;
    }

    bool Piece::operator!=(const Piece& other) const {
        return m_symbol != other.m_symbol;
    }

    Piece Piece::from_color_and_type(char color, char type) {
        if (color == 'w') {
            return Piece(toupper(type));
        } else if (color == 'b') {
            return Piece(tolower(type));
        } else {
            throw new std::invalid_argument("color");
        }
    }

    std::ostream& operator<<(std::ostream& out, const Piece& piece) {
        out << piece.symbol();
        return out;
    }

}
